File: frontend/shader/contract/ts/bitcoin.ts
import { JsonRpcResponse } from "../app/app";

// ============================================================================
// Bitcoin Configuration
// ============================================================================

export interface BitcoinConfig {
  rpcUrl: string;
  rpcUser: string;
  rpcPassword: string;
  network: "mainnet" | "testnet" | "regtest";
}

// ============================================================================
// Bitcoin API Response Types
// ============================================================================

export interface BlockstreamAddressResponse {
  chain_stats?: {
    funded_txo_sum?: number;
  };
  mempool_stats?: {
    funded_txo_sum?: number;
  };
}

/**
 * Bitcoin Connector
 * Connects to Bitcoin blockchain via RPC
 */
export class BitcoinConnector {
  private readonly config: BitcoinConfig;

  constructor(config: BitcoinConfig) {
    this.config = config;
  }

  /**
   * Get balance for a Bitcoin address
   * Uses Bitcoin RPC or public API
   */
  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Bitcoin address");
    }

    try {
      // Try RPC first if configured
      if (
        this.config.rpcUrl &&
        this.config.rpcUser &&
        this.config.rpcPassword
      ) {
        const response = await fetch(this.config.rpcUrl, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Basic ${Buffer.from(
              `${this.config.rpcUser}:${this.config.rpcPassword}`
            ).toString("base64")}`,
          },
          body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "getreceivedbyaddress",
            params: [address, 0],
          }),
        });

        if (response.ok) {
          const data = (await response.json()) as JsonRpcResponse;
          if (data.result !== undefined) {
            // Convert from BTC to satoshis, then back to string
            return (Number.parseFloat(data.result) * 100000000).toString();
          }
        }
      }

      // Fallback to public API (blockchain.info or blockstream.info)
      const apiUrl =
        this.config.network === "mainnet"
          ? `https://blockstream.info/api/address/${address}`
          : `https://blockstream.info/testnet/api/address/${address}`;

      const response = await fetch(apiUrl);
      if (response.ok) {
        const data = (await response.json()) as BlockstreamAddressResponse;
        // Return balance in satoshis
        return data.chain_stats?.funded_txo_sum?.toString() || "0";
      }

      throw new Error("Failed to fetch Bitcoin balance");
    } catch (error) {
      throw new Error(
        `Bitcoin getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  /**
   * Send Bitcoin transaction
   * Note: This requires proper wallet setup and signing
   */
  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string // in satoshis
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress)) {
      throw new Error("Invalid from address");
    }
    if (!this.isValidAddress(toAddress)) {
      throw new Error("Invalid to address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // This requires RPC access with wallet functionality
      if (
        !this.config.rpcUrl ||
        !this.config.rpcUser ||
        !this.config.rpcPassword
      ) {
        throw new Error(
          "Bitcoin RPC credentials required for sending transactions"
        );
      }

      // Get UTXOs for the from address
      const utxoResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 1,
          method: "listunspent",
          params: [0, 9999999, [fromAddress]],
        }),
      });

      if (!utxoResponse.ok) {
        throw new Error("Failed to fetch UTXOs");
      }

      const utxoData = (await utxoResponse.json()) as JsonRpcResponse;
      if (!utxoData.result || (utxoData.result as any[]).length === 0) {
        throw new Error("No UTXOs available for the from address");
      }

      // Create raw transaction
      // Note: This is a simplified version. In production, you'd need proper transaction building
      const createTxResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 2,
          method: "createrawtransaction",
          params: [
            (utxoData.result as any[]).map((utxo: any) => ({
              txid: utxo.txid,
              vout: utxo.vout,
            })),
            { [toAddress]: amountNum / 100000000 }, // Convert satoshis to BTC
          ],
        }),
      });

      if (!createTxResponse.ok) {
        throw new Error("Failed to create raw transaction");
      }

      const createTxData = (await createTxResponse.json()) as JsonRpcResponse;
      const rawTx = createTxData.result as string;

      // Sign transaction
      const signTxResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 3,
          method: "signrawtransactionwithwallet",
          params: [rawTx],
        }),
      });

      if (!signTxResponse.ok) {
        throw new Error("Failed to sign transaction");
      }

      const signTxData = (await signTxResponse.json()) as JsonRpcResponse;
      const signedTx = (signTxData.result as { hex: string }).hex;

      // Send transaction
      const sendTxResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Basic ${Buffer.from(
            `${this.config.rpcUser}:${this.config.rpcPassword}`
          ).toString("base64")}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 4,
          method: "sendrawtransaction",
          params: [signedTx],
        }),
      });

      if (!sendTxResponse.ok) {
        throw new Error("Failed to send transaction");
      }

      const sendTxData = (await sendTxResponse.json()) as JsonRpcResponse;
      return sendTxData.result as string;
    } catch (error) {
      throw new Error(
        `Bitcoin sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  /**
   * Validate Bitcoin address
   */
  isValidAddress(address: string): boolean {
    // Basic validation
    const mainnetPattern =
      /^[13][a-km-zA-HJ-NP-Z1-9]{25,34}$|^bc1[a-z0-9]{39,59}$/;
    const testnetPattern =
      /^[mn2][a-km-zA-HJ-NP-Z1-9]{25,34}$|^tb1[a-z0-9]{39,59}$/;

    if (this.config.network === "mainnet") {
      return mainnetPattern.test(address);
    } else {
      return testnetPattern.test(address) || mainnetPattern.test(address);
    }
  }

  /**
   * Bridge tokens from Bitcoin to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Bitcoin";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["bitcoin", "btc"];
  }
}

File: store/rust/rust/bitcoin.rs
//! Bitcoin connector for interacting with Bitcoin blockchain

use crate::{bitcoin_rpc_call, BitcoinNetwork, BridgeError, BridgeResult};
use serde::{Deserialize, Serialize};

/// Bitcoin connector for interacting with Bitcoin blockchain
pub struct BitcoinConnector {
    rpc_url: String,
    rpc_user: String,
    rpc_password: String,
    network: BitcoinNetwork,
}

impl BitcoinConnector {
    /// Create new Bitcoin connector
    pub fn new(rpc_url: &str, rpc_user: &str, rpc_password: &str, network: BitcoinNetwork) -> BridgeResult<Self> {
        Ok(Self {
            rpc_url: rpc_url.to_string(),
            rpc_user: rpc_user.to_string(),
            rpc_password: rpc_password.to_string(),
            network,
        })
    }

    /// Get balance for a Bitcoin address
    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        // Validate Bitcoin address
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // Get unspent outputs for the address
        let params = serde_json::json!([address]);
        let result = bitcoin_rpc_call(&self.rpc_url, &self.rpc_user, &self.rpc_password, "listunspent", params)?;

        // Sum all unspent outputs
        let utxos: Vec<serde_json::Value> = serde_json::from_value(result)
            .map_err(|e| BridgeError::DeserializationError(format!("Failed to parse UTXOs: {}", e)))?;

        let balance: u64 = utxos
            .iter()
            .filter_map(|utxo| {
                utxo.get("amount")
                    .and_then(|a| a.as_f64())
                    .map(|amt| (amt * 100_000_000.0) as u64) // Convert BTC to satoshis
            })
            .sum();

        Ok(balance)
    }

    /// Send Bitcoin transaction
    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in satoshis
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) {
            return Err(BridgeError::InvalidAddress(from_address.to_string()));
        }
        if !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress(to_address.to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Get UTXOs for from_address
        let params = serde_json::json!([0, 9999999, [from_address]]);
        let result = bitcoin_rpc_call(&self.rpc_url, &self.rpc_user, &self.rpc_password, "listunspent", params)?;
        let utxos: Vec<serde_json::Value> = serde_json::from_value(result)
            .map_err(|e| BridgeError::DeserializationError(format!("Failed to parse UTXOs: {}", e)))?;

        // Calculate total available
        let total_available: u64 = utxos
            .iter()
            .filter_map(|utxo| {
                utxo.get("amount")
                    .and_then(|a| a.as_f64())
                    .map(|amt| (amt * 100_000_000.0) as u64)
            })
            .sum();

        if total_available < amount {
            return Err(BridgeError::InsufficientBalance);
        }

        // Create raw transaction
        // Note: This is a simplified version. In production, you'd need to:
        // 1. Select appropriate UTXOs
        // 2. Calculate fees
        // 3. Create proper transaction structure
        // 4. Sign with private key
        // For now, we'll use sendtoaddress which handles all of this
        let amount_btc = amount as f64 / 100_000_000.0;
        let params = serde_json::json!([to_address, amount_btc]);
        let result = bitcoin_rpc_call(&self.rpc_url, &self.rpc_user, &self.rpc_password, "sendtoaddress", params)?;

        let tx_hash = result
            .as_str()
            .ok_or_else(|| BridgeError::DeserializationError("Invalid transaction hash format".to_string()))?
            .to_string();

        Ok(tx_hash)
    }

    /// Validate Bitcoin address
    pub fn is_valid_address(&self, address: &str) -> bool {
        // Basic validation - in production use proper Bitcoin address validation
        match self.network {
            BitcoinNetwork::Mainnet => {
                address.starts_with('1') || address.starts_with('3') || address.starts_with("bc1")
            },
            BitcoinNetwork::Testnet => {
                address.starts_with('m')
                    || address.starts_with('n')
                    || address.starts_with('2')
                    || address.starts_with("tb1")
            },
            BitcoinNetwork::Regtest => true, // More lenient for regtest
        }
    }

    /// Get transaction info
    pub fn get_transaction(&self, tx_hash: &str) -> BridgeResult<BitcoinTransaction> {
        let params = serde_json::json!([tx_hash, true]);
        let result = bitcoin_rpc_call(&self.rpc_url, &self.rpc_user, &self.rpc_password, "gettransaction", params)?;

        let confirmations = result.get("confirmations").and_then(|c| c.as_u64()).unwrap_or(0) as u32;

        let amount = result
            .get("amount")
            .and_then(|a| a.as_f64())
            .map(|amt| (amt * 100_000_000.0) as i64)
            .unwrap_or(0);

        let fee = result
            .get("fee")
            .and_then(|f| f.as_f64())
            .map(|f| (f.abs() * 100_000_000.0) as u64);

        Ok(BitcoinTransaction {
            txid: tx_hash.to_string(),
            confirmations,
            amount,
            fee,
        })
    }

    /// Get network
    pub fn get_network(&self) -> BitcoinNetwork {
        self.network
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BitcoinTransaction {
    pub txid: String,
    pub confirmations: u32,
    pub amount: i64, // Can be negative for sent transactions
    pub fee: Option<u64>,
}
